File: src/main/java/ltdjms/discord/shared/cache/CacheInvalidationListener.java
package ltdjms.discord.shared.cache;

import java.util.function.Consumer;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import ltdjms.discord.shared.events.BalanceChangedEvent;
import ltdjms.discord.shared.events.DomainEvent;
import ltdjms.discord.shared.events.GameTokenChangedEvent;

/**
 * 緩存失效監聽器。
 *
 * <p>監聽領域事件並使相關緩存失效。 目前為框架層級實作，未來可擴展以實際清除緩存。
 */
public class CacheInvalidationListener implements Consumer<DomainEvent> {

  private static final Logger logger = LoggerFactory.getLogger(CacheInvalidationListener.class);

  private final CacheService cacheService;
  private final CacheKeyGenerator keyGenerator;

  /**
   * 建立緩存失效監聽器。
   *
   * @param cacheService 緩存服務
   * @param keyGenerator 緩存鍵生成器
   */
  public CacheInvalidationListener(CacheService cacheService, CacheKeyGenerator keyGenerator) {
    this.cacheService = cacheService;
    this.keyGenerator = keyGenerator;
  }

  @Override
  public void accept(DomainEvent event) {
    if (event instanceof BalanceChangedEvent e) {
      handleBalanceChanged(e);
    } else if (event instanceof GameTokenChangedEvent e) {
      handleGameTokenChanged(e);
    }
    // 其他事件類型被忽略
  }

  /**
   * 處理貨幣餘額變更事件。
   *
   * @param event 餘額變更事件
   */
  private void handleBalanceChanged(BalanceChangedEvent event) {
    String key = keyGenerator.balanceKey(event.guildId(), event.userId());
    cacheService.invalidate(key);
    logger.debug("使貨幣餘額緩存失效，key: {}", key);
  }

  /**
   * 處理遊戲代幣變更事件。
   *
   * @param event 代幣變更事件
   */
  private void handleGameTokenChanged(GameTokenChangedEvent event) {
    String key = keyGenerator.gameTokenKey(event.guildId(), event.userId());
    cacheService.invalidate(key);
    logger.debug("使遊戲代幣緩存失效，key: {}", key);
  }
}

File: packages/shared/src/infra/cache/cache-invalidation-listener.ts
import { type CacheService } from './cache-service.js';
import { type CacheKeyGenerator } from './cache-key-generator.js';
import { type DomainEvent } from '../../types/events/domain-event.js';
import pino, { type Logger } from 'pino';

/**
 * Listens for economy-related domain events and invalidates the corresponding
 * cache entries so subsequent reads are served from the database.
 *
 * Register this listener via initializeContainer({ eventListeners: [...] }).
 * Gracefully degrades on cache invalidation failure — logs a warning and continues.
 */
export class CacheInvalidationListener {
  private readonly logger: Logger;

  constructor(
    private readonly cacheService: CacheService,
    private readonly cacheKeyGenerator: CacheKeyGenerator,
    logger?: Logger,
  ) {
    this.logger = logger ?? pino({ level: 'silent' });
  }

  /**
   * Handles a domain event and invalidates relevant cache keys.
   * Supported event types:
   *   - balance_changed  → invalidates balanceKey(guildId, userId)
   *   - game_token_changed → invalidates gameTokenKey(guildId, userId)
   * Other event types are silently ignored.
   */
  onEvent(event: DomainEvent): void {
    try {
      switch (event.eventType) {
        case 'balance_changed': {
          const evt = event as unknown as { guildId: string; userId: string };
          if (!evt.guildId || !evt.userId) {
            this.logger.warn({ eventType: event.eventType }, 'Event missing required fields guildId or userId, skipping cache invalidation');
            return;
          }
          const key = this.cacheKeyGenerator.balanceKey(evt.guildId, evt.userId);
          this.cacheService.invalidate(key).catch((err: unknown) => {
            this.logger.warn({ err, key }, 'Failed to invalidate balance cache');
          });
          return;
        }
        case 'game_token_changed': {
          const evt = event as unknown as { guildId: string; userId: string };
          if (!evt.guildId || !evt.userId) {
            this.logger.warn({ eventType: event.eventType }, 'Event missing required fields guildId or userId, skipping cache invalidation');
            return;
          }
          const key = this.cacheKeyGenerator.gameTokenKey(evt.guildId, evt.userId);
          this.cacheService.invalidate(key).catch((err: unknown) => {
            this.logger.warn({ err, key }, 'Failed to invalidate game token cache');
          });
          return;
        }
        default:
          // Unhandled event types are silently ignored
          return;
      }
    } catch (err) {
      this.logger.warn({ err, eventType: event.eventType }, 'Cache invalidation failed');
    }
  }
}
